Add missing TZ conversions and date formatting

th3hamm0r opened the pull request:
**Update:** There is an updated description and testing guide here: https://github.com/wagtail/wagtail/pull/9590#issuecomment-1476379681

Fixes #9581

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

th3hamm0r commented on wagtail/utils/timestamps.py:
```diff
@@ -0,0 +1,11 @@
+from django.utils import formats, timezone
+
+
+def render_timestamp(timestamp):
```
Is it ok to introduce a new module to utils?
Further is the naming ok?

allcaps replied:
Thank you for the work. Great catch!

We might also use:

```
from django.template.defaultfilters import date
date(value, "DATETIME_FORMAT")
```

See also my comment on the issue https://github.com/wagtail/wagtail/issues/9581#issuecomment-1302309601

With this simple syntax, I'm not sure we need a utility method.

Just 2ct. Love the effort!

th3hamm0r replied:
Good point, it would have been nice that way :+1: 
Unfortunately that doesn't work :smile:  https://github.com/wagtail/wagtail/issues/9581#issuecomment-1302315202

laymonage replied:
We already have this exact implementation here:

https://github.com/wagtail/wagtail/blob/9537547b46c4f07938779103c7c42ba424fbe9a6/wagtail/test/utils/timestamps.py#L22-L29

I'm in favour of having this outside of the test module though, so I suppose we can remove that one and change imports to use this one.

th3hamm0r replied:
@laymonage good point, it looks like I've copied it, but also adapted it to use the l18n dependent format instead of the default format (from the settings), which is better for user-faced timestamps. Maybe that's the reason, why I haven't replaced the version from the tests, but I don't know why I did not mention that above :thinking: 

Also I now think there should not be a problem :sweat_smile: 
I will try to replace the test utility with the above one, when I'm rebasing the PR.

th3hamm0r replied:
Speaking about rebasing, #9628 is a tiny fix, which could be reviewed/merged, but it intersects a bit with this PR (tests), so having it merged would help me rebasing this. Sorry for using this thread, but could you do that @laymonage ?

lb- replied:
I think this thread has been resolved.

th3hamm0r replied:
You're right, thx!

th3hamm0r commented:
Update: Just found out, that the timestamps in the logs are not always in the default timezone, some are UTC as well (https://github.com/wagtail/wagtail/issues/9595#issuecomment-1303085511).
We should really add timezone information to the log-timestamps in the long run, or convert them to UTC before storage.

th3hamm0r commented:
I think the only solution to the messed up log timestamps (most of them are stored in UTC, but the "go_live_at"-timestamps are stored with the local timezone), is to migrate those to a consistent format including the timezone information.
This also allows to change the default timezone of the installation, without affecting the output of the stored timestamps.

**Implementation:**
Similar to how django stores the timestamps in the db, and also similar to how django-modelcluster serializes the timestamps in JSON, all timestamps get converted to UTC before writing into the JSONField.
This adds some repeating code snippets, but I've tried to extract most of the code into a new helpers module `wagtail.utils.timestamps`.

I've implemented a migration script that converts the existing log entries based on my findings about their used TZ.

**TBD:**
This change probably requires some notes in the "upgrade" docs, because I've found 3 places, which may be reused by other developers to create similar log entries with a custom "log_action", but which will now use the new format for its dates (they will have to write their own migration script, similar to ours):
* https://github.com/wagtail/wagtail/blob/70cd59dc9a1691aeb2276a6c37c02f3ba5785dca/wagtail/models/__init__.py#L387-L389
* https://github.com/wagtail/wagtail/blob/70cd59dc9a1691aeb2276a6c37c02f3ba5785dca/wagtail/models/__init__.py#L1529-L1531
* https://github.com/wagtail/wagtail/blob/70cd59dc9a1691aeb2276a6c37c02f3ba5785dca/wagtail/actions/publish_revision.py#L177-L179

th3hamm0r commented on wagtail/migrations/0079_fix_log_entry_json_timestamps.py:
```diff
@@ -0,0 +1,129 @@
+            # to UTC by django.
+            item.data["revision"]["created"] = converter(created)
+        except ValueError:
+            # TODO TBD: log error?
```
I was not completely sure, if we should log/fail on ValueErrors (e.g. unexpected date format), or just ignore them. Looking at it again, I now think we should fail on ValueErrors.

As far as I can see, eg. with the introduction of audit logging in https://github.com/wagtail/wagtail/commit/d7d63c3a, the log timestamps have always used the format `%d %b %Y %H:%M`, I haven't found any other format yet. So it should not fail, I think.

th3hamm0r commented on wagtail/migrations/0079_fix_log_entry_json_timestamps.py:
```diff
@@ -0,0 +1,129 @@
+                    go_live_at, tz=timezone.get_default_timezone()
+                )
+        except ValueError:
+            # TODO TBD: log error?
```
Same as above, I think we should not catch the ValueError.

th3hamm0r commented on wagtail/utils/timestamps.py:
```diff
@@ -0,0 +1,41 @@
+            value = timezone.make_aware(value, default_timezone).astimezone(
+                timezone.utc
+            )
+        else:
```
The code is based on django-modelcluster, but think I've found a small improvement compared to the modelcluster's implementation. So I have created a PR there: https://github.com/wagtail/django-modelcluster/pull/168

th3hamm0r commented:
Since the related issue #9581 and the PR itself have been edited and adapted a lot over time, I'm trying to summarize it here again, together with a basic guideline for testing.

**Changes:**

* Replaced all usages of the hardcoded format `%d %b %Y %H:%M` without TZ-conversion, with django's localized version (with formatting depending on the user's language, and converted to the user's configured timezone).
* All timestamps, which get stored in the `BaseLogEntry`'s JSONField, are now stored in UTC, analog to django's handling of db-`DateTimeField`s.
  Previously, depending on the source of the timestamp (`DateTimeField` vs. revision's data), the timestamps have been serialized without any TZ-information in UTC or in the configured default timezone (`settings.TIME_ZONE`).
* All existing timestamps in the JSONField of `BaseLogEntry` are converted with the new migration based on the knowledge, where they initially came from (`DateTimeField` vs. revision's data)
* Changed the `JSONField` "data" in `BaseLogEntry` to use `DjangoJSONEncoder` for serialization, so all new `datetime` instances are serialized automatically with the ISO format. All existing timestamps will be converted to the ISO format with the new migration script.
* Changed tests to use the new `render_timestamp()` utility, removing the usages of the similar test utility `rendered_timestamp()`.

**TODO:**

* Please post your opinion on the above TODOs in the migration script, if we should fail in the migration when invalid date formats are found in those 5 types of log entries. I'm pretty sure, that this should not happen, since I've only found `%d %b %Y %H:%M` in the code. I'm voting for removing the `except ValueError`, so it would interrupt the migration if an invalid format has been found.
* Upgrade notes:
  I think we should note, that the timestamps in the audit log's data are now all stored in UTC, and the migration updates all core log entries to the new ISO format. If the developer uses its own logs with extra JSON data, there are new utils in `wagtail.utils.timestamps` to handle the time conversions.
  
  I've also found 3 places, where a developer may use different action names for "wagtail.revert" and "wagtail.publish". If the developer uses different strings, those log entries won't be converted by the migration script (as it only handles the 5 known actions). So the developer must create his own migration, based on ours. But I'm not sure, because those APIs are mainly documented in the code and in case of `wagtail.actions.publish_revision.PublishRevisionAction` it is even noted in the doc strings, that you cannot pass a string (but one could still do it, I think).
    * https://github.com/wagtail/wagtail/blob/dd77f5b32fd2c2353eda9dd857b527a8f9d1fa08/wagtail/models/__init__.py#L392-L394
    * https://github.com/wagtail/wagtail/blob/dd77f5b32fd2c2353eda9dd857b527a8f9d1fa08/wagtail/models/__init__.py#L1764-L1766
    * https://github.com/wagtail/wagtail/blob/dd77f5b32fd2c2353eda9dd857b527a8f9d1fa08/wagtail/actions/publish_revision.py#L187-L189


**Testing:**

I'd recommend, that you first install a current version of wagtail in a virtualenv (but don't use 4.1.3/4.2.1 due to #10243), configure the default `TIME_ZONE` (which is usually either set to the project/server's local time or UTC) with some offset to your local timezone, but don't use UTC to better see all issues, and familiarize yourself with the issues listed below. Further, if you change the language to a non-english language, you will also notice that the date formatting is not in the way expected for the language.

Example steps:
1) `pip install wagtail && wagtail start mysite && cd mysite && pip install -r requirements.txt && ./manage.py makemigrations && ./manage.py migrate && ./manage.py createsuperuser`
2) In `mysite/mysite/settings/base.py` change the `TIME_ZONE` to something with some offset to your local timezone (not UTC), for me in "Europe/Vienna" I could set it to `TIME_ZONE = "Australia/Brisbane"`.
3) `./manage.py runserver`
4) Login to the CMS
5) Go to the account settings and change the current timezone to your local timezone, and optionally the preferred language to a non-english language, which has different date formatting.
6) Now, add, change, publish, revert, schedule a page, cancel the schedule,... and always try to remember at which time/hour you did this, or which time/hour you've used for the "Go live at" date fields. You will notice, that all notifications and all audit log entries which contain timestamps, will not show the dates in your expected local time (and formatting, if you've changed the preferred language). Instead you will find 2 types of buggy timestamps:
   * Some dates will be in UTC:
     Basically all dates, which are directly retrieved from a django-`DateTimeField`, because django stores and retrieves all dates in UTC, and the old handling just formats it with `%d %b %Y %H:%M`. This also includes those dates, which have been retrieved from such a field, and then stored into the audit log.
   * Some dates will be in the configured `TIME_ZONE`:
     You will find these dates in the audit log for those entries, which contain 2 timestamps. One is the creation timestamp (which is UTC), and the other one is the `go_live_at` timestamp (which uses `TIME_ZONE`). The later one has been retrieved from the revisions JSON data initially (and converted to the `TIME_ZONE`), before it has been stored in the audit log's data.
7) Now, let's install the fixed version of wagtail and run the fixing migration. For that you will have to checkout the source of this PR somewhere locally (see [wagtail's docs](https://docs.wagtail.org/en/latest/contributing/developing.html) and [github's docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally)) and then install it with something like:
   `pip install -e ~/YOUR/LOCATION/OF/WAGTAIL && ./manage.py migrate`
8) You should now find all timestamps in notifications and the audit log in your local timezone, and with correct formatting for the selected language.


Overview of fixed issues:

1) "Persistent" issues (only the migration in combination with the code changes fix those issues):
   On the writing side (loggings), search for the usage of `ensure_utc()`.
   On the reading side (log listings), search for the combined usage of `parse_datetime_localized()` and `render_timestamp()`.
   
   These issues happen for all audit log entries, which use the JSON "data" field to store timestamps:

    * Log entries with only one timestamp `created` (previously stored in UTC):
        * `wagtail.revert`
        * `wagtail.rename`
        * `wagtail.publish`

    * Log entries with 2 timestamps, `created` (previously stored in UTC) and `go_live_at` (previously stored in the default timezone, `TIME_ZONE`):
        * `wagtail.publish.schedule`
        * `wagtail.schedule.cancel`
    

2) "Non-persistent" issues (fixed without the migration):
   Basically search for the other usages of `render_timestamp()`
   
   Those issues mainly happen in notification/lock messages. Since those views/mixins are often generic mixins, you should find those issues with non-page models too:

    * Review a different revision (the warning message at the top should display the correct time)
    * Revert a page to another revision (save-only)
    * Revert a page to another revision (publish)
    * Revert a page to another revision (scheduled publication)
    * Open a scheduled (locked) page

lb- commented:
Haven't had a chance to run locally but the code all makes sense and appears to be solid. I'll do my best to run through your testing instructions tomorrow.

laymonage commented on wagtail/migrations/0084_fix_log_entry_json_timestamps.py:
```diff
@@ -0,0 +1,129 @@
+def migrate_schedule_logs(model, converter):
+    for item in model.objects.filter(
+        action__in=["wagtail.publish.schedule", "wagtail.schedule.cancel"]
+    ).only("data"):
```
Use `.iterator()` in case of really big websites with lots of logs?

th3hamm0r replied:
Nice catch, I think I've used both in a project, but I will look into this tomorrow 👍

th3hamm0r replied:
fixed

laymonage reviewed (commented):
Haven't had the chance to test and give a proper review, but something caught my eye

laymonage commented on wagtail/migrations/0084_fix_log_entry_json_timestamps.py:
```diff
@@ -0,0 +1,137 @@
+            item.save(update_fields=["data"])
+
+
+def migrate_custom_to_iso_format(apps, schema_editor):
+    ModelLogEntry = apps.get_model("wagtailcore.ModelLogEntry")
+    PageLogEntry = apps.get_model("wagtailcore.PageLogEntry")
+
+    migrate_logs_with_created_only(ModelLogEntry, legacy_to_iso_format)
+    migrate_logs_with_created_only(PageLogEntry, legacy_to_iso_format)
+
+    migrate_schedule_logs(ModelLogEntry, legacy_to_iso_format)
+    migrate_schedule_logs(PageLogEntry, legacy_to_iso_format)
```
Note that we support [custom `FooLogEntry` classes](https://docs.wagtail.org/en/stable/extending/audit_log.html#log-models). I'm not sure if we should try to migrate those logs as well or leave it to the developers (at least with a mention in the upgrade considerations, linking to this migration file). I don't know if it's even possible to do so though, as the log model registry uses real models and I don't know if we can get the fake model from there to use in migrations.

th3hamm0r replied:
Yes, loading those models without loading the real models looks impossible to me.
I think we can only mention this in the upgrade notes, maybe together with the undocumented way of using different action names mentioned in the todo-notes above.

th3hamm0r commented:
@lb- I've tried hard to reproduce this with your TZ settings, but I wasn't able to.
The only thing I can explain to me is, if the `TIME_ZONE` setting is changed BEFORE executing the new migration, the migration script will convert the existing timestamps incorrectly, but that's nothing we can fix. Actually, the PR fixes it, by using UTC for all timestamps. Afterwards, changing `TIME_ZONE` should not be an issue anymore (as with any django `DateTimeField` for example).

Based on that, we could of course mention somewhere (e.g. in docs for older versions), that for any wagtail version before this PR, changing the `TIME_ZONE` will break the audit-log timestamps, so must be avoided.

lb- commented:
Thanks @th3hamm0r  for the additional steps to get this tested, sorry I did not get a chance to do this earlier. I am still keen to understand and review this PR though but it may not be for another little while as I am about to go away. Requesting myself for another review and putting notes here.


1.  Install the bakerydemo, but don't start runserver yet: https://github.com/wagtail/bakerydemo#setup-with-virtualenv
2.  In bakerydemo/settings/base.py change the TIME_ZONE = "Europe/Vienna".
3. ./manage.py runserver
4. Login to the CMS (admin:changeme)
5. Go to the account settings http://127.0.0.1:8000/admin/account/ and change the current time zone to "Australia/Brisbane"
6. Add a new test page:
 * http://127.0.0.1:8000/admin/pages/add/base/standardpage/60/
 * Insert a title
 * In the status panel, click "Set schedule"
 * Configure a "Go live date/time" in the future, let's take "2023-04-26 09:00"
 * click "Save schedule"
 * "Publish" the page
7. Re-open the edit page view for the scheduled page: http://127.0.0.1:8000/admin/pages/84/edit/
  The lock notification should now display "25 Apr 2023 23:00" (= scheduled date in UTC)), instead of "April 26, 2023, 9 a.m.". Not only is the time incorrect, but also the format is not consistent with the other places in the admin. The status bar should display the correct "Go-live" date ("April 26, 2023, 9 a.m.").
8. Open the page history view: http://127.0.0.1:8000/admin/pages/84/history/
  The top line should display "Page scheduled for publishing at 26 Apr 2023 01:00" (= scheduled date in Europe/Vienna), instead of "Page scheduled for publishing at April 26, 2023, 9 a.m.".
9. Stop the runserver
10. Now, install the fixed version of wagtail (ensure you've rebuilt the frontend assets) and run the fixing migration: `pip install -e ~/YOUR/LOCATION/OF/WAGTAIL && ./manage.py migrate`
11. The history view should now display the correct date and format ("April 26, 2023, 9 a.m.").
12. The edit view's lock notification should now display the correct date and format too.

lb- commented:
Sorry I haven't reviewed this yet, I actually saw the issue you've described while testing the bakerydemo Squash build today. 

It was saying my last edit on a page was 10hrs ago, even though I'd just made it. 

Would be great if you can take a look though @allcaps

lb- reviewed (approved):
@th3hamm0r finally sat down to do this in full again and it works great, thank you!

I'll rebase and merge this in shortly.

## Screenshots - before
Observe - set to **01/08/2023 at 10:00am**
<img width="900" alt="Screenshot 2023-07-13 at 7 28 23 am" src="https://github.com/wagtail/wagtail/assets/1396140/2c0aa218-40e4-4ec1-b254-bfc7d133719d">

Observe - Times are incorrectly shown on the messaging and audit history.

<img width="1268" alt="Screenshot 2023-07-13 at 7 29 31 am" src="https://github.com/wagtail/wagtail/assets/1396140/bffe9b91-9849-4057-a8d7-1cd2067c862b">
<img width="1133" alt="Screenshot 2023-07-13 at 7 28 58 am" src="https://github.com/wagtail/wagtail/assets/1396140/3a8bc43b-ec47-423c-9f4e-8e30f41f26f0">

## Screenshots - after

Observe, times are now shown correctly.

<img width="1270" alt="Screenshot 2023-07-13 at 7 32 41 am" src="https://github.com/wagtail/wagtail/assets/1396140/c209d555-d048-4909-ad32-e94e72a0ea36">

<img width="1153" alt="Screenshot 2023-07-13 at 7 32 47 am" src="https://github.com/wagtail/wagtail/assets/1396140/dbff7564-b206-42d6-8e3f-6fc7ce9c27b1">

lb- commented:
@th3hamm0r  I think we should lean towards a 'safer' migration that does not error so let's leave this as is.

Can you please prepare another PR that just adds the logging in instead?

`# TODO TBD: log error?`

Also, can you include the upgrade considerations as noted in this other PR.

th3hamm0r commented:
Thanks a lot @lb- !
I'll create a PR shortly.

lb- commented:
Thanks @th3hamm0r

laymonage commented on wagtail/migrations/0088_fix_log_entry_json_timestamps.py:
```diff
@@ -0,0 +1,137 @@
+            # If a previous_revision was available, the data contains "revision" with
+            # its created date.
+            # Also, there are "wagtail.publish" logs, which don't set data at all.
+            created = item.data["revision"]["created"]
+            # "created" is set to the previous revision's created_at, which is set
+            # to UTC by django.
+            item.data["revision"]["created"] = converter(created)
+        except ValueError:
+            # TODO TBD: log error?
+            continue
+        except KeyError:
+            continue
```
I tried running this migration on torchbox.com's Wagtail instance and it failed with a different error: `TypeError: 'NoneType' object is not subscriptable`

The likely cause was that there are old logs where `log.data` is stored as a JSON `null` (not SQL `NULL` or an empty JSON object `{}`) in the database, and it's deserialized into `None` in Python, so accessing `data["revision"]` will fail. More on that in https://github.com/wagtail/wagtail/pull/8021#issuecomment-1658082683.

I think we should guard against `TypeError` here to avoid this. It would also avoid a potentially unlikely issue where `data["revision"]` is `None`.
